libsolidity/codegen/ir/Common.cpp

```diff
@@ -16,13 +16,28 @@
 */
 
 #include <libsolidity/codegen/ir/Common.h>
+#include <libsolidity/ast/TypeProvider.h>
 
 #include <libsolutil/CommonIO.h>
 
 using namespace std;
 using namespace solidity::util;
 using namespace solidity::frontend;
 
+Arity Arity::fromDefinition(FunctionDefinition const& _function)
+{
+	FunctionType const* functionType = TypeProvider::function(_function)->asCallableFunction(false);
```

Comment:
I'm actually not so sure about the `asCallableFunction` here.

Which arity of the function are you actually talking about here? This should always be the yul-arity when calling the function internally, right?

Reply:
I took `asCallableFunction()` from `internalDispatch()`. I only moved it to this function. Actually, its use was not removed from `internalDispatch()` until #8797 so it might not be easy to see. I just moved this change here for clarity. See the updated diff.

As for sol vs yul: it's yul-arity everywhere since that's what matters to internal dispatch. But can yul-arity even be different than sol-arity? At least for functions that are not bound (bound ones cannot be stored in pointers anyway)? If it can then yeah, might be a good idea to add `yul` somewhere in the function name...

Reply:
Some Solidity variables take two stack slots. It probably was a bug in the old code. What you need here is
``TypeProvider::function(_function, FunctionType::Kind::Internal)``.

Reply:
Looks like it's not really a bug. After looking into this, I see that `asCallableFunction()` converts `calldata` arguments to `memory` arguments which could change the number of slots for dynamic arrays. But as far as I understand, only `external` functions can have `calldata` arguments and these never go through internal dispatch.

So I think it's just superfluous, not buggy. I created a test that confirms this. It's in a separate PR since it's only tangentially related to the refactor here: #8984.

Reply:
While it is true that calldata can only be used for external functions now, this is about to change very soon: https://github.com/ethereum/solidity/pull/8942

Reply:
Good to know. Would it actually affect this code though? `asCallableFunction()` will have to be updated not to replace `calldata` parameters with `memory` ones, right?

Reply:
Why would you want to call `asCallableFunction` at all?

Reply:
I wouldn't. Just wondering if the code was actually broken or just doing something unnecessary that did not and would not actually affect the result. Well, it's not that important though. If the new version is the right way to do it then it's all good now.